### PR TITLE
Finish the introduction of the new wires logic

### LINKED ISSUE #1030: [BUG] Broadcasting multi-qubit gates in tape mode produces new duplicate wires
#### Issue description

When using the broadcast function in tape mode, broadcasting multi-qubit gates seems to create new wires that have type tensor.

* *Expected behavior:* Broadcasted gates should be applied on the device wires.

* *Actual behavior:* Additional wires are created that have type tensor.

* *Reproduces how often:* Whenever multi-qubit gates are included in the broadcast. 

* *System information:* 
(Using dev version from 2020-01-24)

```
Name: PennyLane
Version: 0.14.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/olivia/Software/anaconda3/envs/xanadu/lib/python3.8/site-packages
Requires: scipy, networkx, semantic-version, autograd, toml, numpy, appdirs
Required-by: PennyLane-SF, pennylane-qulacs, PennyLane-qsharp, PennyLane-qiskit, PennyLane-Qchem, PennyLane-Forest, PennyLane-Cirq
Platform info:           Linux-5.4.0-64-generic-x86_64-with-glibc2.10
Python version:          3.8.5
Numpy version:           1.19.5
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.14.0.dev0)
- default.mixed (PennyLane-0.14.0.dev0)
- default.qubit (PennyLane-0.14.0.dev0)
- default.qubit.autograd (PennyLane-0.14.0.dev0)
- default.qubit.jax (PennyLane-0.14.0.dev0)
- default.qubit.tf (PennyLane-0.14.0.dev0)
- default.tensor (PennyLane-0.14.0.dev0)
- default.tensor.tf (PennyLane-0.14.0.dev0)
- strawberryfields.fock (PennyLane-SF-0.11.0)
- strawberryfields.gaussian (PennyLane-SF-0.11.0)
- strawberryfields.remote (PennyLane-SF-0.11.0)
- qulacs.simulator (pennylane-qulacs-0.12.0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.8.0)
- qiskit.aer (PennyLane-qiskit-0.13.0.dev0)
- qiskit.basicaer (PennyLane-qiskit-0.13.0.dev0)
- qiskit.ibmq (PennyLane-qiskit-0.13.0.dev0)
- forest.numpy_wavefunction (PennyLane-Forest-0.11.0)
- forest.qpu (PennyLane-Forest-0.11.0)
- forest.qvm (PennyLane-Forest-0.11.0)
- forest.wavefunction (PennyLane-Forest-0.11.0)
- cirq.mixedsimulator (PennyLane-Cirq-0.13.0)
- cirq.pasqal (PennyLane-Cirq-0.13.0)
- cirq.qsim (PennyLane-Cirq-0.13.0)
- cirq.qsimh (PennyLane-Cirq-0.13.0)
- cirq.simulator (PennyLane-Cirq-0.13.0)


```

#### Source code and tracebacks

```
import pennylane as qml
import numpy as np
qml.enable_tape()

N_states = 4
params = np.random.uniform(low=-np.pi/2, high=np.pi/2, size=(N_states))

def circuit(params, wires):
    qml.broadcast(qml.RY, wires, pattern='single', parameters=params)
    return qml.expval(qml.PauliZ(0))

dev = qml.device("default.qubit", wires=N_states)
my_qnode = qml.QNode(circuit, dev)
my_qnode(params, range(N_states))
print(my_qnode.draw())
```
This produces the expected output:

```
 0: ──RY(-0.195)──┤ ⟨Z⟩ 
 1: ──RY(-0.18)───┤     
 2: ──RY(-0.779)──┤     
 3: ──RY(0.404)───┤     
```

However altering the function to include a multi-qubit gate, like so:
```
def circuit(params, wires):
    qml.broadcast(qml.RY, wires, pattern='single', parameters=params)
    qml.broadcast(qml.CRY, wires, pattern='ring', parameters=params)
    return qml.expval(qml.PauliZ(0))
```
gives the following output after running and drawing the tape:
```
 0: ──RY(-0.21)──╭C─────────────────────────────────╭RY(1.14)──┤ ⟨Z⟩ 
 1: ──RY(1.38)───╰RY(-0.21)──╭C─────────────────────│──────────┤     
 2: ──RY(-1.05)──────────────╰RY(1.38)──╭C──────────│──────────┤     
 3: ──RY(1.14)──────────────────────────╰RY(-1.05)──╰C─────────┤     
 0: ──RY(-0.21)───C──────────────────────────────────RY(1.14)──┤ ⟨Z⟩ 
 1: ──RY(1.38)────RY(-0.21)───C────────────────────────────────┤     
 2: ──RY(-1.05)───────────────RY(1.38)───C─────────────────────┤     
 3: ──RY(1.14)───────────────────────────RY(-1.05)───C─────────┤     

```
It doesn't seem to be an issue just in the tape drawer, because looking at the tape operations shows that the CRY are applied to different wires:
```
[RY(tensor(-0.20956558, requires_grad=True), wires=[0]),
 RY(tensor(1.37704322, requires_grad=True), wires=[1]),
 RY(tensor(-1.05444866, requires_grad=True), wires=[2]),
 RY(tensor(1.14254618, requires_grad=True), wires=[3]),
 CRY(tensor(-0.20956558, requires_grad=True), wires=[tensor(0, requires_grad=True), tensor(1, requires_grad=True)]),
 CRY(tensor(1.37704322, requires_grad=True), wires=[tensor(1, requires_grad=True), tensor(2, requires_grad=True)]),
 CRY(tensor(-1.05444866, requires_grad=True), wires=[tensor(2, requires_grad=True), tensor(3, requires_grad=True)]),
 CRY(tensor(1.14254618, requires_grad=True), wires=[tensor(3, requires_grad=True), tensor(0, requires_grad=True)])]
```

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**
We had a lot of PRs that change the wires class to make it more efficient. Amongst them, the logic of how the `Wires` class interprets the inputs has been changed to be cleaner and more streamlined.

This has introduced a potential bug stemming from the fact that iterables are exposed to different checks (i.e. they can be anything hashable) than non-iterables (which can only be numbers or strings). The problem is that a user may use an iterable of wire labels to create wires on the device, but cannot address them with a single object since the single object does not pass the stricter check.

This PR resolves this issue and fixes the wires logic once and for all: if we can iterate over an input, the elements returned by iteration are interpreted as the labels, else the input is interpreted as a label. Labels have to be hashable. 

**Changes:**
* Rewrites the `_process` function and updates the changelog to present all changes, as well as some docstrings. 
* Removes some stuff in default.qubit.autograd and tf that converts 0-dim tensor wires but causes nasty side effects. 
* Fixes one small bug in the broadcast function, while I was at it.

**Benefits:**
One can now define in a few sentences how the wires class will treat its inputs (see changelog). This is consistent and logical. The PR will also fix #1030.

**Possible Drawbacks:**

While all tests pass, this PR introduces one last substantial change: The ever-problematic case of `Wires(np.array(1))` was previously unpacked and interpreted as a wire label `1`. Now `np.array(1)` is interpreted as the label, and an error is raised because it is unhashable. Note that `pennylane.numpy` arrays _are_ hashable, so this will not be a problem when wires are accidentally (or by our current positional-arguments-are-trainable design) turned into tensors. I think this is the right behaviour though, because `Wires([np.array(1)])` will raise an error on current master as well, as an iterable with non-hashable elements.

### CODE REVIEW

File: tests/interfaces/test_autograd.py [hunk 1]
```diff
@@ -893,7 +893,8 @@ def circuit(x, y, z):
     def test_non_diff_wires_argument(self, w, expected_res, expected_grad, tol):
         """Test that passing wires as a non-differentiable positional
         argument works correctly."""
-        dev = qml.device("default.qubit", wires=2)
+        dev = qml.device("default.qubit", wires=[qml.numpy.array(0, requires_grad=False),
+                                                 qml.numpy.array(1, requires_grad=False)])
```
Comment (mariaschuld) [inline tests/interfaces/test_autograd.py:897]: This test relies on the fact that in autograd we currently convert positional arguments to `pennylane.numpy` arrays to be backwards compatible.

File: pennylane/wires.py [hunk 2]
```diff
@@ -25,55 +24,71 @@ class WireError(Exception):
...
+    unique and hashable. Else it is interpreted as a single wire label that has to be hashable. The
+    only exception are strings which are interpreted as wire labels.
 
-    There is no conceptual difference between registers of multiple wires and single wires,
-    which are just wire registers of length one.
+    The hash function of a wire label is considered the source of truth when deciding whether
+    two wire labels are the same or not.
 
-    Indexing and slicing this sequence will return another ``Wires`` object.
+    Indexing an instance of this class will return a wire label.
```
Comment (mariaschuld) [inline pennylane/wires.py:88]: We forgot to update this docstring previously...

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -340,14 +340,37 @@
...
+    
+  - Indexing wires objects now returns a label, instead of a new `Wires` object. For example:
+    
+    ```pycon
+    >>> w = Wires([0, 1, 2])
+    >>> w[1]
+    >>> 1
+    ```
+     
+  - The check for uniqueness of wires moved from `Wires` instantiation to
```
Comment (mariaschuld) [inline .github/CHANGELOG.md:367]: The changelog was a bit thin anyways...

File: pennylane/devices/default_qubit_autograd.py [hunk 4]
```diff
@@ -169,17 +169,3 @@ def _get_unitary_matrix(self, unitary):
...
-        Args:
-            wires (Wires): wires whose labels we want to map to the device's internal labelling scheme
-
-        Returns:
-            Wires: wires with new labels
-        """
-        wires = Wires([w.item() if isinstance(w, tensor) else w for w in wires])
-        mapped_wires = super().map_wires(wires)
-
-        return mapped_wires
```
Comment (mariaschuld) [inline pennylane/devices/default_qubit_autograd.py:185]: I think that this was a fix someone introduced because of our inconsistent treatment of the 0-dim tensor. I remove it because it totally disturbs the (now sound) wires logic.

File: pennylane/templates/broadcast.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def wires_all_to_all(wires):
 
 # define wire sequences for patterns
 PATTERN_TO_WIRES = {
-    "single": lambda wires: [wires[i] for i in range(len(wires))],
+    "single": lambda wires: [wires.subset([i]) for i in range(len(wires))],
```
Comment (mariaschuld) [inline pennylane/templates/broadcast.py:68]: Snuck this in here: the change of indexing makes this pattern behave differently from multi-wire ones. Should be safe and helps to fix #1030.

**Comment by mariaschuld** [comment]
Ignore codecov error @Thenerdstation, the number of misses did not increase (only the number of hit lines decreased from deleting code). Ready for review :)

File: pennylane/templates/broadcast.py (hunk 5 above)
Comment (Thenerdstation) [reply pennylane/templates/broadcast.py:68]: 👍

File: pennylane/wires.py [hunk 6]
```diff
@@ -25,55 +24,71 @@ class WireError(Exception):
...
-            set_of_wires = set(wires)
+    except TypeError:
+        # if not iterable, interpret as single wire label
+        try:
+            hash(wires)
         except TypeError as e:
-            # Make sure it really was a hashability issue
+            # if object is not hashable, cannot identify unique wires
             if str(e).startswith("unhashable"):
                 raise WireError("Wires must be hashable; got {}.".format(wires)) from e
```
Comment (Thenerdstation) [inline pennylane/wires.py:60]: Maybe?

```suggestion
                raise WireError("Wires must be hashable; got object of type {}.".format(type(wires))) from e
```

**Review by Thenerdstation — APPROVED** [review]
I think this is a positive change. Certainly better than the crazy amount of type checking we had before.

File: pennylane/wires.py [hunk 7]
```diff
@@ -25,55 +24,71 @@ class WireError(Exception):
...
+        # This is the only exception to the logic
+        # of considering the elements of iterables as wire labels.
+        wires = [wires]
 
-    if isinstance(wires, Iterable):
+    try:
+        # Use tuple conversion as a check for whether `wires` can be iterated over.
+        # Note, this is not the same as `isinstance(wires, Iterable)` which would
+        # pass for 0-dim numpy arrays that cannot be iterated over.
         tuple_of_wires = tuple(wires)
```
Comment (albi3ro) [inline pennylane/wires.py:52]: Some I'm confused about what `tuple` would catch that `set` wouldn't.

**Comment by albi3ro** [comment]
I looked over everything, and it all looked good to me :)
Hooray for efficiency!

File: pennylane/devices/default_qubit.py [hunk 8]
```diff
@@ -149,7 +150,15 @@ def __init__(self, wires, *, shots=1000, analytic=True, cache=0):
     def map_wires(self, wires):
         # temporarily overwrite this method to bypass
         # wire map that produces Wires objects
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:152]: Perhaps better to add more details on what needs to be done long term here? E.g., `# TODO:...`

File: pennylane/devices/default_qubit_autograd.py (hunk 4 above)
Comment (josh146) [reply pennylane/devices/default_qubit_autograd.py:185]: This was probably me 😆

File: pennylane/templates/broadcast.py (hunk 5 above)
Comment (josh146) [reply pennylane/templates/broadcast.py:68]: 🔥

File: tests/interfaces/test_autograd.py (hunk 1 above)
Comment (josh146) [reply tests/interfaces/test_autograd.py:897]: Nice! And this test verifies that #1030 is solved, from what I can tell?

**Review by josh146 — APPROVED** [review]

File: pennylane/wires.py (hunk 7 above)
Comment (mariaschuld) [reply pennylane/wires.py:52]: Oh, it may also be able to decide iterability, but we need a tuple anyways...

File: tests/interfaces/test_autograd.py (hunk 1 above)
Comment (mariaschuld) [reply tests/interfaces/test_autograd.py:897]: Yes, I tried and it works. @glassnotes is also happy with this solution.
